### PR TITLE
DataDictParallel

### PR DESCRIPTION
**Pull request by tkornut**
In this pull request I am adding DataDictParallel, a specialized wrapper class derived from PyTorch DataParallel class that can handle DataDicts.

It enables to parallelize processing of batches on many GPUs by deploying model replicas and scattering batches between them.

The approach works for feed-forward models, including ConvNets and TorchVisionWrappers, also when several models in the pipeline.

Example: CNN + FFN Softmax Classifier on CPU:

Screen Shot 2019-05-02 at 10 49 28 PM

The same with 1 GPU:

Screen Shot 2019-05-02 at 10 50 08 PM

And 5 GPUs:

Screen Shot 2019-05-02 at 10 55 14 PM

GPU utilization when training image-only based model (VGG16 + FFN Softmax Classifier) on VQAMED C2:

Screen Shot 2019-05-02 at 10 54 41 PM

Sadly, there are some issues, mostly with RNNs. Seems that it is not correctly performing scattering (i.e. there are 5 batches of 64 gathered at the end resulting in batch of size 320 instead of 64):

Screen Shot 2019-05-02 at 11 25 07 PM

For that reason for now it is optional, activated by the flag *--dataparallel*

When those issues will be solved, the flag will be removed.

### CODE REVIEW

**Comment by tkornut** [comment]
I have identified the source of error in scattering of strings. 
This results in SentenceEmbeddings returning all samples for all devices, which, after gathering, results in batch of size = input batch size * num devices.

The solution for now is the list of components that *won't be wrapped in DataDictParallell" by definition. Tested the solution on many pipelines, seem to work.

Example: pipeline with > 30 components, using 6 GPUs:
Screen Shot 2019-05-03 at 9 38 53 AM

LSTM + VGG16 + FFNs using all of them:
Screen Shot 2019-05-03 at 9 39 51 AM

INFO: Making the parallel mode default from now on - when you use more than I GPU, the computations *will be parallelized*! Simply add --gpu and that's it!

**Comment by tkornut** [comment]
**INFO:

Making the parallel mode default from now on - when you use more than I GPU, the computations will be parallelized! Simply add *--gpu* when running ptp worker and that's it!**
